pandas/core/construction.py

```diff
@@ -329,7 +337,8 @@ def array(
     if dtype is None:
         inferred_dtype = lib.infer_dtype(data, skipna=True)
         if inferred_dtype == "period":
-            return PeriodArray._from_sequence(data, copy=copy)
+            period_data = cast(Union[Sequence[Optional[Period]], AnyArrayLike], data)
```

Comment:
wouldn't mind a mypy ignore instead of the cast

Reply:
Do you prefer the ignore? I personally think the cast is more semantically correct, since we checked the type using `lib.infer_dtype`. This pattern also seems to be used in different parts of the codebase already, too ( [example](https://github.com/pandas-dev/pandas/blob/main/pandas/core/apply.py#L217)).

Reply:
Do you prefer the ignore? I personally think the cast is more semantically correct, since we checked the type using `lib.infer_dtype`. This pattern also seems to be used in different parts of the codebase already, too ( [example](https://github.com/pandas-dev/pandas/blob/main/pandas/core/apply.py#L217)).

Reply:
I have no strong opinion (slightly in favor of ignore, unless it creates many follow up ignores) @simonjayhawkins has a better overview when to use `cast`s I am/was a bit too liberal using them :)

Reply:
we have checks for redundant casts and also unused ignores so if, in general, the typing issue is likely to be fixed in the future then either are OK (but I personally prefer an ignore since we traditionally add the mypy error as a comment, so when looking at the code, the error message is perhaps more helpful than a cast)

In this case, where the issue is that a type if not correctly narrowed by a condition then I am equally happy with a cast adjacent to the condition.